python/ray/data/_internal/execution/operators/map_transformer.py

```diff
@@ -284,14 +417,14 @@ def create_map_transformer_from_block_fn(
 class RowMapTransformFn(MapTransformFn):
     """A rows-to-rows MapTransformFn."""
 
-    def __init__(self, row_fn: MapTransformCallable[Row, Row], is_udf: bool = False):
-        self._row_fn = row_fn
+    def __init__(self, udf_context: MapTransformUDFContext):
         super().__init__(
             MapTransformFnDataType.Row,
             MapTransformFnDataType.Row,
             category=MapTransformFnCategory.DataProcess,
-            is_udf=is_udf,
+            udf_context=udf_context,
         )
+        self._row_fn = self._generate_transform_fn_from_udf_context()
```

Comment:
We need to defer this until the `init` method.
otherwise, the `_row_fn` may capture a different `self`. 
I guess this PR should be working with this change. 
But the current design is a bit too convoluted and can be simplified. 
I'll prototype something and get back.